extensions/amp-viewer-integration/0.1/amp-viewer-integration.js

```diff
@@ -14,5 +14,88 @@
  * limitations under the License.
  */
 
+import {Messaging} from './messaging.js';
+import {listen} from '../../../src/event-helper';
+import {viewerForDoc} from '../../../src/viewer';
+import {dev,user} from '../../../src/log';
+
+
+/**
+ * @fileoverview This is the communication protocol between AMP and the viewer.
+ * This should be included in an AMP html file to communicate with the viewer.
+ */
 export class AmpViewerIntegration {
+
+  /**
+   * @param {!Window} win
+   */
+  constructor(win) {
+    /** @const @private {!Window} win */
+    this.win_ = win;
+  }
+
+  /**
+   * Initiate the handshake. If handshake confirmed, start listening for
+   * messages.
+   * @return {!Promise|undefined}
+   */
+  init() {
+    const isIframed = window.parent && window.parent != window;
+    if (!isIframed) {
+      return;
+    }
+    dev().info('AVI', 'amp-viewer-integration.js => handshake init()');
+    const viewer = viewerForDoc(this.win_.document);
+    return this.getHandshakePromise_(viewer)
+      .then(viewerOrigin => {
+        dev().info('AVI',
+          'amp-viewer-integration.js => listening for messages');
+        const messaging_ =
+          new Messaging(this.win_, this.win_.parent, viewerOrigin,
+            (type, payload, awaitResponse) => {
+              return viewer.receiveMessage(
+                type, /** @type {!JSONType} */ (payload), awaitResponse);
+            });
+        viewer.setMessageDeliverer((type, payload, awaitResponse) => {
+          return messaging_.sendRequest(type, payload, awaitResponse);
+        }, viewerOrigin);
+      });
+  }
+
+  /**
+   * Send a handshake request, and listen for a handshake response to
+   * confirm the handshake.
+   * @param {!../../../src/service/viewer-impl.Viewer} viewer
+   * @return {!Promise}
+   * @private
+   */
+  getHandshakePromise_(viewer) {
+    const win = this.win_;
+
+    return new Promise(resolve => {
+      const unconfirmedViewerOrigin = viewer.getParam('viewerorigin');
+      user().assert(unconfirmedViewerOrigin,
+              'Expected viewer origin must be specified!');
+
+      const unlisten = listen(win, 'message', listener);
+      function listener(event) {
+        if (event.origin == unconfirmedViewerOrigin &&
+            event.data == 'amp-handshake-response' &&
+            event.source == win.parent) {
+          dev().info('AVI', 'amp-viewer-integration.js => received handshake ' +
+            'confirmation');
+          unlisten();
```

Comment:
Can probably wait until later, but what I'm concerned about is that we may have a risk condition here: viewer may immediately start sending messages after issuing handshake response, but we will miss these messages in the time between `unlisten` and the next `listen` later. I think we can delay thinking about this until we start reconciling with PROD API, but in the meantime `TODO` is warranted.